core/src/main/scala/kafka/server/checkpoints/CheckpointFile.scala

```diff
@@ -58,11 +62,8 @@ class CheckpointFile[T](val file: File, version: Int, formatter: CheckpointFileF
         fileOutputStream.getFD().sync()
       } catch {
         case e: FileNotFoundException =>
-          if (FileSystems.getDefault.isReadOnly) {
-            fatal(s"Halting writes to checkpoint file (${file.getAbsolutePath}) because the underlying file system is inaccessible: ", e)
-            Exit.halt(1)
-          }
-          throw e
+          logDirFailureChannel.maybeAddLogFailureEvent(file.getParentFile.getParent)
+          throw new KafkaStorageException(s"Halting writes to checkpoint file (${file.getAbsolutePath}) because the underlying file system is inaccessible", e)
```

Comment:
I have thought about catching IOException here. After reading https://github.com/apache/kafka/pull/698/files I am still not sure why we previously didn't halt the Kafka on any IOException here. There are also some IOException that are explicitly thrown in this class that do not trigger halt.

I couldn't find a reason not to catch IOException either. I choose to be conservative by preserving the behavior by not doing this on all IOException just in case we have a reason not to do so.

Would you like me to do this on any IOException? Alternatively, maybe we can discuss the handling of these specific IOException in a followup patch?